src/DeviceInput/InputDevices/deviceSourceManager.ts

```diff
@@ -0,0 +1,308 @@
+import { DeviceInputSystem, POINTER_DEVICE, KEYBOARD_DEVICE, MOUSE_DEVICE } from '../deviceInputSystem';
+import { Engine } from '../../Engines/engine';
+import { IDisposable } from '../../scene';
+import { DeviceType, DeviceInputs } from './deviceEnums';
+import { Nullable } from '../../types';
+import { Observable } from '../../Misc/observable';
+
+/**
+ * Class that handles all input for a specific device
+ */
+export class DeviceSource<T extends DeviceType> {
+    /**
+     * Observable to handle device input changes per device
+     */
+    public onInputChangedObservable = new Observable<{ inputIndex: number, previousState: Nullable<number>, currentState: Nullable<number> }>();
+    private _deviceInputSystem: DeviceInputSystem;
+    private _touchPoints: Array<string>;
+
+    /**
+     * Default Constructor
+     * @param deviceInputSystem Reference to DeviceInputSystem
+     * @param deviceName Name of device to be used by DeviceInputSystem
+     * @param deviceType Type of device
+     * @param deviceSlot "Slot" or index that device is referenced in
+     */
+    constructor(deviceInputSystem: DeviceInputSystem,
+        /** Name of device to be used by DeviceInputSystem */
+        public deviceName: string,
+        /** Type of device */
+        public deviceType: DeviceInputs<T>,
+        /** "Slot" or index that device is referenced in */
+        public deviceSlot: number = 0) {
+        this._deviceInputSystem = deviceInputSystem;
+
+        if (deviceType == DeviceType.Touch) {
+            this._touchPoints = new Array<string>();
+        }
+    }
+
+    /**
+     * Get input for specific input
+     * @param inputIndex index of specific input on device
+     * @returns Input value from DeviceInputSystem
+     */
+    public getInput(inputIndex: T): Nullable<number> {
+        if (this.deviceType == DeviceType.Touch) {
+            return this._deviceInputSystem.pollInput(this._touchPoints[this.deviceSlot], inputIndex);
+        }
+
+        return this._deviceInputSystem.pollInput(this.deviceName, inputIndex);
+    }
+
+    /**
+     * Add specific point to array of touch points (DeviceType.Touch only)
+     * @param name Name of Specific Touch Point
+     */
+    public addTouchPoints(name: string) {
+        if (this.deviceType == DeviceType.Touch) {
+            this._touchPoints.push(name);
+        }
+    }
+
+    /**
+     * Remove specific point from array of touch points (DeviceType.Touch only)
+     * @param name Name of Specific Touch Point
+     */
+    public removeTouchPoints(name: string) {
+        if (this.deviceType == DeviceType.Touch) {
+            const touchIndex = this._touchPoints.indexOf(name);
+            this._touchPoints.splice(touchIndex, 1);
+        }
+    }

```

Comment:
I think these are only publicly exposed so that `DeviceSourceManager` can set them. This is a bit different from what I had imagined. I thought all the state would still be managed by `DeviceSourceManager`, and `DeviceSource` would just provide a typesafe way to interact with a specific device managed by `DeviceSourceManager`. I think doing that would make a bit more sense, and also probably making `DeviceSource` a nested class of `DeviceSourceManager` so that `DeviceSource` can call private methods of `DeviceSourceManager`. This way the public contract will be very clear and not provide access to private state.

Reply:
My initial idea was to just keep track of how many touch inputs there were and keep them in a manageable set of slots so that the user doesn't need to keep up with a bunch of changing pointer ids.  For the time being, I've ripped out all of that code and will be testing things without it.